### PR TITLE
feat(profiling): Enable profiling for ASGI frameworks

### PR DESCRIPTION
**Pull request by Zylphrex**
This enables profiling for ASGI frameworks. When running in ASGI sync views, the
transaction gets started in the main thread then the request is dispatched to a
handler thread. We want to set the handler thread as the active thread id to
ensure that profiles will show it on first render.

### CODE REVIEW

File: sentry_sdk/profiler.py [hunk 1]
```diff
@@ -305,13 +304,22 @@ def __init__(
...
 
     def __enter__(self):
         # type: () -> None
+        hub = self.hub or sentry_sdk.Hub.current
+
+        _, scope = hub._stack[-1]
+        old_profile = scope.profile
+        scope.profile = self
+
+        self._context_manager_state = (hub, scope, old_profile)
```
Comment (sl0thentr0py) [inline sentry_sdk/profiler.py:321]: maybe cleaner to do `configure_scope/scope.profile = profile` in the `start_profiling` ctx manager below? Then you don't need this extra state.

File: sentry_sdk/integrations/fastapi.py [hunk 2]
```diff
@@ -62,6 +65,22 @@ def patch_get_request_handler():
 
     def _sentry_get_request_handler(*args, **kwargs):
         # type: (*Any, **Any) -> Any
+        dependant = kwargs.get("dependant")
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/fastapi.py:68]: @antonpirker can you pls review the fastapi part?

File: sentry_sdk/integrations/quart.py [hunk 3]
```diff
@@ -71,18 +77,62 @@ def setup_once():
         got_request_exception.connect(_capture_exception)
         got_websocket_exception.connect(_capture_exception)
 
-        old_app = Quart.__call__
+        patch_asgi_app()
+        patch_scaffold_route()
+
+
+def patch_asgi_app():
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/quart.py:84]: can you split quart out into a separate PR? We can get the quart maintainer to take a look as well then so he's aware of the changes.

File: sentry_sdk/profiler.py (hunk 1 above)
Comment (Zylphrex) [reply sentry_sdk/profiler.py:329]: That context manager is going to be replaced in an upcoming change to enable profiling for all transactions since I need to hook into the transaction which does not allow me to use the context manager.

File: sentry_sdk/integrations/quart.py (hunk 3 above)
Comment (Zylphrex) [reply sentry_sdk/integrations/quart.py]: https://github.com/getsentry/sentry-python/pull/1830

File: sentry_sdk/integrations/fastapi.py [hunk 4]
```diff
@@ -62,6 +65,22 @@ def patch_get_request_handler():
...
+        if dependant and not asyncio.iscoroutinefunction(dependant.call):
+            old_call = dependant.call
```
Comment (antonpirker) [inline sentry_sdk/integrations/fastapi.py:70]: It seems that `dependant.call` could be `None`: https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/models.py#L27

Do we have to check for that here to prevent calling None as a function dow at line 80?

File: sentry_sdk/integrations/fastapi.py [hunk 5]
```diff
@@ -62,6 +65,22 @@ def patch_get_request_handler():
...
+
+            def _sentry_call(*args, **kwargs):
+                # type: (*Any, **Any) -> Any
+                hub = Hub.current
+                with hub.configure_scope() as sentry_scope:
+                    if sentry_scope.profile is not None:
+                        sentry_scope.profile.active_thread_id = (
+                            threading.current_thread().ident
```
Comment (antonpirker) [inline sentry_sdk/integrations/fastapi.py:78]: Because `current_thread()` can be a dummy object (see https://docs.python.org/3/library/threading.html#threading.current_thread) maybe it is saver to use ` threading.get_ident()`?
(But if the dummy object also has a `.ident` my comment can be ignored)

Comment (Zylphrex) [reply sentry_sdk/integrations/fastapi.py:78]: `_DummyThread` inherits from `Thread` ([source](https://github.com/python/cpython/blob/main/Lib/threading.py#L1437)) and so both should have the `ident` [property](https://docs.python.org/3/library/threading.html#threading.Thread.ident).

File: sentry_sdk/integrations/fastapi.py (hunk 4 above)
Comment (Zylphrex) [reply sentry_sdk/integrations/fastapi.py:74]: Good catch! Updated!

**Review by sl0thentr0py — APPROVED** [review]
